aws/s3_test.go

```diff
@@ -458,3 +462,103 @@ func TestNukeS3Bucket(t *testing.T) {
 		})
 	}
 }
+
+// TestFilterS3BucketArgs represents arguments for TestFilterS3Bucket_Config
+type TestFilterS3BucketArgs struct {
+	configFilePath string
+	matches        []string
+}
+
+func bucketNamesForConfigTests() []string {
+	return []string{
+		"alb-alb-123456-access-logs-" + S3TestGenBucketName(),
+		"alb-alb-234567-access-logs-" + S3TestGenBucketName(),
+		"tonico-prod-alb-access-logs-" + S3TestGenBucketName(),
+		"prod-alb-public-access-logs-" + S3TestGenBucketName(),
+		"stage-alb-internal-access-logs-" + S3TestGenBucketName(),
+		"stage-alb-public-access-logs-" + S3TestGenBucketName(),
+		"cloud-watch-logs-staging-" + S3TestGenBucketName(),
+		"something-else-logs-staging-" + S3TestGenBucketName(),
+	}
+}
+
+// TestFilterS3Bucket_Config tests listing only S3 buckets that match config file
+func TestFilterS3Bucket_Config(t *testing.T) {
+	t.Parallel()
+
+	// Create test buckets
+	awsParams, err := newS3TestAWSParams()
+	require.NoError(t, err, "Failed to setup AWS params")
+
+	var bucketTags []map[string]string
+	bucketNames := bucketNamesForConfigTests()
+	for _, bucketName := range bucketNames {
+		err = S3TestCreateBucket(awsParams.svc, bucketName, bucketTags, false)
+		require.NoErrorf(t, err, "Failed to create test bucket - %s", bucketName)
+	}
+
+	awsSession, err := S3TestCreateNewAWSSession("")
+	require.NoError(t, err, "Failed to create random session")
+
+	// Define test cases
+	type testCaseStruct struct {
+		name string
+		args TestFilterS3BucketArgs
+	}
+
+	includeBuckets := []string{}
+	includeBuckets = append(includeBuckets, bucketNames[:4]...)
+
+	excludeBuckets := []string{}
+	excludeBuckets = append(excludeBuckets, bucketNames[:3]...)
+	excludeBuckets = append(excludeBuckets, bucketNames[4])
+	excludeBuckets = append(excludeBuckets, bucketNames[6:]...)
+
+	filterBuckets := []string{}
+	filterBuckets = append(filterBuckets, bucketNames[:3]...)
+
+	testCases := []testCaseStruct{
+		{
+			"Include",
+			TestFilterS3BucketArgs{
+				configFilePath: "../config/mocks/s3_include_names.yaml",
+				matches:        includeBuckets,
+			},
+		},
+		{
+			"Exclude",
+			TestFilterS3BucketArgs{
+				configFilePath: "../config/mocks/s3_exclude_names.yaml",
+				matches:        excludeBuckets,
+			},
+		},
+		{
+			"IncludeAndExclude",
+			TestFilterS3BucketArgs{
+				configFilePath: "../config/mocks/s3_filter_names.yaml",
+				matches:        filterBuckets,
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		// Capture the range variable as per https://blog.golang.org/subtests
+		// Not doing this will lead to tc being set to the last entry in the testCases
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			// Clean up test buckets
+			defer nukeAllS3Buckets(awsParams.awsSession, aws.StringSlice(bucketNames), 1000)
```

Comment:
Doesn't this introduce a race condition where depending on how fast the tests go, one test could nuke buckets needed for another test? Also, could this be "wasteful" in that we call nuke multiple times?

I'm guessing that this works fine when running this test in isolation, but I wonder if our CI build will encounter the race condition when it is running all the tests and has many things going on, causing more context switching.

Consider using the test group pattern to run `nukeAllS3Buckets` once when all the parallel tests finish:

```go
defer nukeAllS3Buckets()
t.Run("group", func(t *testing.T) {
  for _, tc := range testCases {
    tc := tc
    t.Run(tc.name, func(t *testing.T) {
      t.Parallel()
      // test code
    })
  }
})
```

Reply:
> I wonder if our CI build will encounter the race condition when it is running all the tests and has many things going on, causing more context switching.

It looks like it is not surefire to run into this given the most recent test run, but I think it is still in the realm of possibility that it might be worth adjusting this, either to the group model or adjusting the names list so that it only affects the buckets for that test.

Reply:
Should that `defer` be inside of the outer `t.Run`? Isn't the problem that it will nuke the buckets before the tests run?

Reply:
What you have does work, but how? How is it that when I had a t.Run, just a single one in the for loop, with defer further up (just after the bucketNames were defined), it ran defer before the tests? But here it is still outside of a t.Run (two, in fact) and somehow runs after the tests as it should?

I guess the key is that `defer` somehow knows to wait for `t.Run` to return. When `t.Run` executes within a `for` loop, defer doesn't wait, unless the `for` loop is wrapped in a `t.Run`. Is each `t.Run` in the loop run in parallel, is that how `t.Parallel` works? The way the test-case code inside the innermost `t.Run` is written, it looks lexically like it's just running a single test-case in parallel (which doesn't make sense). But maybe `t.Parallel` somehow knows that the outer `for` loop it's in defines all the stuff that should run in parallel? If that's the case, then maybe `defer` can't know that those multiple `t.Run`s in the `for` loop are done, and _that's_ why we need to wrap the loop in another `t.Run`, which is easy for `defer` to wait for.... ???

I'll research myself, but those are my thoughts trying to make sense of how this works.